source/extensions/filters/http/jwt_authn/filter_config.h

```diff
@@ -55,48 +55,56 @@ struct JwtAuthnFilterStats {
 };
 
 /**
- * The filter config object to hold config and relevant objects.
+ * The filter config interface. It is an interface so that we can mock it in tests.
  */
-class FilterConfig : public Logger::Loggable<Logger::Id::jwt>, public AuthFactory {
+class FilterConfig {
 public:
-  ~FilterConfig() override = default;
+  virtual ~FilterConfig() = default;
 
-  FilterConfig(envoy::extensions::filters::http::jwt_authn::v3::JwtAuthentication proto_config,
-               const std::string& stats_prefix, Server::Configuration::FactoryContext& context)
-      : proto_config_(std::move(proto_config)),
-        stats_(generateStats(stats_prefix, context.scope())),
-        tls_(context.threadLocal().allocateSlot()), cm_(context.clusterManager()),
-        time_source_(context.dispatcher().timeSource()), api_(context.api()) {
-    ENVOY_LOG(info, "Loaded JwtAuthConfig: {}", proto_config_.DebugString());
-    tls_->set([this](Event::Dispatcher&) -> ThreadLocal::ThreadLocalObjectSharedPtr {
-      return std::make_shared<ThreadLocalCache>(proto_config_, time_source_, api_);
-    });
+  virtual JwtAuthnFilterStats& stats() PURE;
 
-    for (const auto& rule : proto_config_.rules()) {
-      rule_pairs_.emplace_back(Matcher::create(rule),
-                               Verifier::create(rule.requires(), proto_config_.providers(), *this));
-    }
+  virtual bool bypassCorsPreflightRequest() const PURE;
 
-    if (proto_config_.has_filter_state_rules()) {
-      filter_state_name_ = proto_config_.filter_state_rules().name();
-      for (const auto& it : proto_config_.filter_state_rules().requires()) {
-        filter_state_verifiers_.emplace(
-            it.first, Verifier::create(it.second, proto_config_.providers(), *this));
-      }
-    }
-  }
+  // Finds the matcher that matched the header
+  virtual const Verifier* findVerifier(const Http::HeaderMap& headers,
+                                       const StreamInfo::FilterState& filter_state) const PURE;
+};
+using FilterConfigSharedPtr = std::shared_ptr<FilterConfig>;
 
-  JwtAuthnFilterStats& stats() { return stats_; }
+/**
+ * The filter config object to hold config and relevant objects.
+ */
+class FilterConfigImpl : public Logger::Loggable<Logger::Id::jwt>,
+                         public FilterConfig,
+                         public AuthFactory,
+                         public std::enable_shared_from_this<FilterConfigImpl> {
+public:
+  ~FilterConfigImpl() override = default;
+
+  // Finds the matcher that matched the header
+  static std::shared_ptr<FilterConfigImpl>
+  create(envoy::extensions::filters::http::jwt_authn::v3::JwtAuthentication proto_config,
+         const std::string& stats_prefix, Server::Configuration::FactoryContext& context) {
+    std::shared_ptr<FilterConfigImpl> ptr(
+        new FilterConfigImpl(proto_config, stats_prefix, context));
```

Comment:
nit: use make_shared here

Reply:
i can't use make shared because the constructor is private

Reply:
added a comment